### DISCUSSION TITLE
May release 0.9 - Introducing a new Budibase (Beta)

### DISCUSSION THREAD

**Discussion by shogunpurple** [Budibase/budibase#1557] [Announcements]
# May release 0.9 - Introducing a new Budibase

The May release transforms the development experience for every single Budibase user. It is also the beginning of a new and exciting journey that involves Budibase moving to the browser.

### TL; DR

*   Run Budibase via the browser in your self-hosted cloud
*   We will no longer actively develop on the Electron Budibase builder
*   A new Admin UI
*   Improved user management
*   Built-in email support
*   Google Auth - users can sign in with Google
*   New, fresh, accessible UI
*   New data provider component (with sorting/filtering/UX/perf enhancements)
*   New table component (remove and add columns)
*   New formula data type 
*   Import Options as part of the CSV import
*   Add code to automations
*   CRON automation triggers

### Run Budibase in the cloud

From this release forward, you will run Budibase via the browser in your self-hosted cloud. We will cease development on our old friend, the Budibase desktop builder (Electron). Electron served us well, but the cloud is where our future lies for a number of reasons:

*   Flexibility - users can now access their Budibase account on multiple devices.
*   Collaboration - this is our first major step towards multi-user collaboration.
*   Faster releases - developing on Electron brought with it additional work and effort that slowed us down. When developing in the cloud, we can ship new features quicker.

Getting started with the new Budibase

To run Budibase in the cloud, you will need Docker or Digital Ocean. The setup is straightforward - especially for Digital Ocean. You can access the docs [here](http://docs.budibase.com/).

_We hope to add more cloud services over the upcoming months, and welcome help/contributions from the community._

What is going to happen to old Budibase and my apps

The old Budibase will remain as it is today. We will no longer develop and enhance it. 

Old apps developed in the old Budibase (Electron) can be downloaded and uploaded to your new Budibase instance in your self-hosted cloud environment.

---

### New Admin UI

Previous to this release, there was very little administrative functionality. This was not scalable, so we’ve created a new UI dedicated to administrative tasks such as end-user management, developer management, email/SMTP, and authentication. This UI is only available to the admin of your Budibase account.

More importantly, this admin interface defines our direction - a platform where companies build multiple apps that empower their workforce, automate processes, and accelerate digital transformation. 

---

### Improved user management

The process of adding, updated and removing users is getting centralized. So, all you need to do is add a user to the platform once, then assign them to different apps. This is incredibly helpful for you and end-user  - end-users only have to sign in once, to access multiple apps!

With this release, when adding a user you have two options. If you have SMTP set-up, you can add them via email (more on that below), or add them manually. 

---

### Built-in email and templates

Onboarding new users is always a tricky task and a pain to set up. Thankfully, with this version of Budibase, all you have to do is add your SMTP settings and Budibase will take care of the onboarding process for you. So, once you add a user, Budibase will know to send them a welcome email and ask them to set up a password. And, for users who forget their password, Budibase has an email workflow for that too. \
 \
The email templates, such as, forgot password, are editable templates that you can access in the new admin panel.  \

---

### Google Auth

You can now add Google Auth to your apps in minutes. This will allow your users to sign in to your apps with their Google credentials.

_We plan on adding additional auth options. [Please let us know if you have a particular requirement.](https://github.com/Budibase/budibase/discussions/categories/ideas)_

----

### New, fresh, and accessible builder UI

With the 0.8 release, we added new form components to make it easier for users to build better, accessible forms. To do this, we used Adobe’s Spectrum CSS components. The project was a huge success, and we’ve decided to use Spectrum for our own components which are used to create the actual Budibase builder UI. Out of the box, users will have 4 themes to choose from; lightest, light, dark, darkest. All components are accessibility tested, and support RTL alignment (not launched yet). 

We believe the builder is looking better than ever and welcome all feedback.

---

### New data provider component 

The data provider component makes it easier and faster to build data-rich UIs. It is similar to the repeater component but with some additional special features.

The new data provider component consolidates the fetching of data to one component. So, if you are creating a dashboard with 2 charts and a table (pulling data from the same source), you would only need to use one data provider component. Previously you would have to use 4 repeaters, which is a worse user experience and less performant.

With the new data provider component, you can also filter and sort the data you pull in from the data section. For example, if you have a list of holiday homes stored within your database, you can pull that data into your design section, and filter/sort that data by location, all using the data provider component. Additionally, the data provider component allows you to customize the amount of data rows you want to display and paginate your data with just one click. It’s immensely powerful.

---

### New table component (remove and add columns)

The previous table component was powerful but lacked flexibility and was not performant. The new table is lightweight, flexible to use, and themeable. Most importantly, once you pull in data from your data sources, you can now add and remove tables columns - yay! 

Additionally, elements of the table are now exposed in the nav tree, allowing you to add components to your tables.

---

### Formulas

You can now add a formula data type to Budibase’s internal database. This will allow you to make calculations based on data within your internal tables. You can also use the formula data type to ‘pluck’ data from another internal table and use that data in a calculation in your present table.

---

### CSV import update - upload options!

Users can now upload a CSV to Budibase, and select whether that CSVs columns are the following data types: Text, Number, Date, and now, Options. For many users who are upgrading their Excel / Google sheets to Budibase, this is a huge time saver and we hope you enjoy it.

---

### Add code to automations

To increase the utility of automations, we have added a new scripting block, allowing you to add JavaScript to your automations. This exponentially increases the extensibility of automations.

---

### CRON automation trigger

You can now schedule your automations. For example - Run automation every 5 minutes. This is particularly good for sending reports.

---

**Comment by KernOil** [comment]
migrating from the windows BB builder to the docker web based builder, all of my design screens are blank. is there a fix for this?

**Reply by shogunpurple** [reply]
Hi @AndrewJRamay - would you be able to let us know if there are any errors in your dev console when you visit the design screen? I tried to reproduce with the app you sent over and while it did take quite a while to load, it did work successfully. An error log from your budibase setup would really help us debug.

**Reply by KernOil** [reply]
ok let me check

**Reply by KernOil** [reply]
ok this is what i am getting.
![image](https://user-images.githubusercontent.com/80058238/119546981-60d89100-bd49-11eb-9d97-d55baf91acce.png)

**Reply by shogunpurple** [reply]
Thanks @AndrewJRamay. There's a few things you can try:
- Make sure you have definitely pulled the latest docker images. Run `budi hosting --update`
- Import your app again after this
- Try clearing your browser cookies, or running in incognito. There may be overhang from the old version of budibase.

**Reply by KernOil** [reply]
ok ill delete everything and try it again! BRB

**Reply by KernOil** [reply]
I uninstalled everything and am trying to get it installed again this is where im getting stuck now any ideas 
![image](https://user-images.githubusercontent.com/80058238/119565383-b9feef80-bd5e-11eb-9c44-044f976478aa.png)

**Reply by mike12345567** [reply]
Hi @AndrewJRamay - two things:
1.  Does the `.env` file exist where you initialised Budibase? It may be that there is an environment file in your home directory already, I would recommend creating a directory to initialise Budibase in, e.g. `mkdir ~/budibase` and then re-running the `budi hosting --init` in the directory, you shouldn't need `sudo` to do this.
2. On Linux you may need to setup permissions for docker/docker-compose, if this is the issue you'll still experience the error if you just run `docker-compose up` in the directory you initialised Budibase. You can find info about this [here](https://www.digitalocean.com/community/questions/how-to-fix-docker-got-permission-denied-while-trying-to-connect-to-the-docker-daemon-socket)

**Reply by KernOil** [reply]
Thank you I will give that a shot!

**Reply by KernOil** [reply]
the re installation worked got it back up and running 
![image](https://user-images.githubusercontent.com/80058238/119569547-ba4db980-bd63-11eb-9f9d-537eedd4013a.png)


but no change on the builder errors 

![image](https://user-images.githubusercontent.com/80058238/119569600-ce91b680-bd63-11eb-8a0d-e037146054fc.png)

**Reply by mike12345567** [reply]
Hi @AndrewJRamay can you create a new app out of interest, just to check its a problem with the imported file or something not setup correctly on the platform.

**Reply by KernOil** [reply]
I did try that. created a new app and am having the same problem with the new app as with the older imported one. I'm assuming it has to be my installation I just have no idea what part tho.

**Reply by shogunpurple** [reply]
That path seems to be outdated - `app-assets`. It is now `prod-budi-app-assets`. Did you do a `budi hosting --update` before running?

**Reply by KernOil** [reply]
last week I installed the old bb docker container just to practice. I uninstalled all the docker containers today and  ran the setup for the new bb container.  I just ran the updater. has a few changes. the error is different now.
![image](https://user-images.githubusercontent.com/80058238/119571235-f550ec80-bd65-11eb-8cee-b9c0b8dab416.png)

**Reply by shogunpurple** [reply]
Ok, awesome. So now try re-importing your app, and see if that works. Think this might be the final step 🤞

**Reply by KernOil** [reply]
ok will do that now!

**Reply by KernOil** [reply]
still has that same error but I can see the web pages now! Thank you very much!

![image](https://user-images.githubusercontent.com/80058238/119571903-c9823680-bd66-11eb-9c5b-7c3b9836b11c.png)

**Reply by shogunpurple** [reply]
Think that might just be from the last time - if you refresh it should go away I think. Great! Glad you are up and running 👍

**Reply by KernOil** [reply]
a refresh did it, no more errors. 
Thank you for hanging in there with me. I really like your platform I hope this is a big success for you.

**Comment by KernOil** [comment]
I have found a bug in the new release. the button function shows no query's 
![image](https://user-images.githubusercontent.com/80058238/119576408-8bd4dc00-bd6d-11eb-8b30-3187b1b1fdf7.png)

it should be showing these two.
![image](https://user-images.githubusercontent.com/80058238/119576456-a14a0600-bd6d-11eb-80ce-b3ccef092b2b.png)

**Reply by aptkingston** [reply]
You're right @AndrewJRamay, that is indeed a bug! I'll get this fixed ASAP.

**Reply by aptkingston** [reply]
Should be fixed via https://github.com/Budibase/budibase/pull/1563 👍

**Reply by KernOil** [reply]
Thank you sir. Do I just run "budi hosting --update" for the update?

**Comment by KernOil** [comment]
how do users work in the new 0.9 release. it looks like I still have my old users table in my app but there is also a users section in the admin consul. However I can not get either of them to work for the credentials to log into my imported app. 
am I doing something wrong?

**Reply by mjashanks** [reply]
Users should be added and remove through the admin section. The users table in the app allows you to control roles for the user in that particular app (which you can also do in the admin section).

We need the user's table in the app so that we can do relationships to users. e.g. "Created By".

We should not import users into apps if they have not been added to the admin section. Your old credentials will not get imported - you will need to create your users again, via the admin section.

**Reply by KernOil** [reply]
gotcha, I have added the user in the admin consul, I can not add anything more then email. I then add the same email and user info + password in the app's table 'users'. When I try and use those credentials to log into the app (Email / password ) I get the response forbidden even trying both passwords one generated in the admin consul and the user defined password in the app's table.
what am I doing wrong?     also there is no longer a log out button in the top right of the app not sure if that was intentional or not.

**Reply by mike12345567** [reply]
> gotcha, I have added the user in the admin consul, I can not add anything more then email. I then add the same email and user info + password in the app's table 'users'. When I try and use those credentials to log into the app (Email / password ) I get the response forbidden even trying both passwords one generated in the admin consul and the user defined password in the app's table.
> what am I doing wrong? also there is no longer a log out button in the top right of the app not sure if that was intentional or not.

Sorry @AndrewJRamay - there was a couple of bugs around this which have now been fixed, you can pull down the latest changes with `budi hosting --update` - that'll update your system and fix the forbidden errors when logging in.

The log out button was removed from the apps as it didn't make sense anymore, as you can't login to an app directly (you log into the Budibase platform itself) logging out through an app would have been a bit confusing (as this would log you out of the platform itself).

You log out now through the portal itself, as such:
![image](https://user-images.githubusercontent.com/4407001/119861637-844d3880-bf0f-11eb-82b0-68642c6eb633.png)

**Reply by KernOil** [reply]
Thank you for the clarification. ill update and play with it more. 
BTW I am loving the new update role out speed able to fix things in real time that is awesome!

**Reply by KernOil** [reply]
@mike12345567  So is this log in screen dead then I can not point my URL here and have my users use this login screen anymore?
![image](https://user-images.githubusercontent.com/80058238/119868299-b8a30380-bed3-11eb-8f41-109c9f388c8f.png)

**Reply by KernOil** [reply]
@mike12345567 After updating I can not get past the above screen to preview my app. what  credentials should be used here?

**Reply by aptkingston** [reply]
@AndrewJRamay that login screen is now defunct and it no longer comes with new apps. You should delete that screen, and users will be automatically logged in to every app once they authenticate to the platform. So for any old apps, the migration strategy is just to delete the login screen.

**Reply by KernOil** [reply]
perfect thank you

**Reply by KernOil** [reply]
@aptkingston i deleted my public screen and i have no screens in the public folder. however the above log in screen is now replaced with this log in screen when trying to preview the app before publishing it.
![image](https://user-images.githubusercontent.com/80058238/119874898-fb1c0e80-beda-11eb-89f2-1e1dc021098a.png)

**Comment by KernOil** [comment]
Is there a way I can use a URL link in a button? Right now it seams like I can only use it  with a link element. and that element makes my app freeze when used.

**Reply by aptkingston** [reply]
If you're trying to link to an external URL you'll need to tick the `External` setting on link component, in case that is making it freeze. And yes, you can use a button. There is a button action for "Navigate To" which takes a URL.

**Reply by KernOil** [reply]
thank you for the tip on the link having to check the external box. that has fixed my freezing problem. I still can not get the button to work with an external link using the navigate to feature. when clicking on it nothing happens.

**Comment by KernOil** [comment]
if I have a basic user with only 1 app why dose it go to this screen first? why can it not just jump into the app?
![image](https://user-images.githubusercontent.com/80058238/119897702-aafe7580-bef5-11eb-9337-aca00e3f7f04.png)

**Reply by joebudi** [reply]
Hey Andrew. This is something we have chatted about internally and is something we may implement in the future.

The idea with the portal, is many users build multiple apps for different depts within their workplace. And once you build more than a couple of apps, it gets hard to manage the links. 

Also, for example, in some cases you may only want your marketing dept to see the marketing apps, and HR to see only the apps relevant to them.

**Reply by KernOil** [reply]
ok, yes I totally agree with multiple apps that makes sense, I think it would be more streamlined for non admen users who only have one app, have the option to skip that screen and log into their single app.  also is there a way to customize the log in screen like in version 0.8?

**Reply by joebudi** [reply]
How would you like to customize it?

**Reply by KernOil** [reply]
it would be adding our company logo and name to this screen 
![image](https://user-images.githubusercontent.com/80058238/119899483-fdd92c80-bef7-11eb-8fad-a779a6341716.png)

like i did here 
![image](https://user-images.githubusercontent.com/80058238/119868299-b8a30380-bed3-11eb-8f41-109c9f388c8f.png)

**Reply by joebudi** [reply]
This makes sense. I'll bring it to the team tomorrow.

**Reply by KernOil** [reply]
that would be awesome thank you.

**Reply by KernOil** [reply]
> This makes sense. I'll bring it to the team tomorrow.

for my app the old design worked better with logging in and out. the new design is much harder for users to be switched. my users will be using the same computer for their reporting under different user accounts, and will be logging in and out every hour 24 hours a day.   a quicker in and out would work best in my application.

**Reply by KernOil** [reply]
> Hey Andrew. This is something we have chatted about internally and is something we may implement in the future.
> 
> The idea with the portal, is many users build multiple apps for different depts within their workplace. And once you build more than a couple of apps, it gets hard to manage the links.
> 
> Also, for example, in some cases you may only want your marketing dept to see the marketing apps, and HR to see only the apps relevant to them.

Hi @joebudi thank you for the quick implementation of stream lining the log in of single app users, a huge improvement! The problem I am having now is there is no way to log that user out once in the app. can you add a log out button I can put in my app to log that user out. when I close out that screen the web browser auto logs that user back in.

**Comment by KernOil** [comment]
Hi Guys, I am finding a lot of the data connections in automatons do not work in this new update, along with a few other things. 

1) I can not get a web hook to fire i keep getting this error
![image](https://user-images.githubusercontent.com/80058238/120003657-cf586180-bf8a-11eb-9f39-851fba1eb87f.png)

2) The trigger external database query dos not have any of my actual connections in it it is only showing old connections, 
![image](https://user-images.githubusercontent.com/80058238/120003839-f7e05b80-bf8a-11eb-9707-262606ad8b74.png)

3) When submitting a form I can not leave the date field blank without it erroring out the query, it will save internally but wont write to my PostgreSQL DB externally  
![image](https://user-images.githubusercontent.com/80058238/120004092-383fd980-bf8b-11eb-8a77-388be9843917.png)

4) There also seams to be a problem with value limits and required for example: I can not submit my form with these fields blank. 
![image](https://user-images.githubusercontent.com/80058238/120004549-c2883d80-bf8b-11eb-95bc-ae91615a75e7.png)

as you can see here I am not requiring a value for the column so I should not be stopped by a null value.  
![image](https://user-images.githubusercontent.com/80058238/120004709-f19eaf00-bf8b-11eb-9a31-93fedf7db566.png)
also if I hit submit and it tells me i can not submit because these values cant be null, i then input the given values and submit. it will save to internal DB but it will fail the external DB query  

5) I can not duplicate a button in the design tab I have a fair amount of things going on with the click and would like to have two buttons with just minor navigation differences. cant get it to duplicate like the other items. 
![image](https://user-images.githubusercontent.com/80058238/120005516-c072ae80-bf8c-11eb-9076-aa97c7af8b77.png)

**Reply by mike12345567** [reply]
Hi @AndrewJRamay - on the first point about the webhook trigger, could you provide a bit more information about this? For example the automation you've created that you can't get to trigger - is this part of an old app which has been imported?

**Comment by Manishkparmar80** [comment]
hi, Google Auth is added into new version.  

Most of organization having Active directory configured and all users having domain username and password for login into computer.  

**How we can incorporate LDAP Configuration for SSO (Single Sign On) functionality into Budibase portal for domain username and password based authentication for APP instead of manual username & password of local database of Budibase.**

**Reply by mjashanks** [reply]
Hi There. This is not available right now - but is on our roadmap. We hope to have this over the next few months.

**Reply by Manishkparmar80** [reply]
Thanks for confirmation. Do i need to put feature request so that it gets track properly and we may get faster delivery of this feature.

**Comment by Manishkparmar80** [comment]
How we can add 2- factor authentication by integrating with tool like Google Authenticator.

**Reply by mjashanks** [reply]
Our built-in authentication (email and password) does not support 2fa. As above, we will be supporting other auth methods, which will cover 2fa.

**Reply by Manishkparmar80** [reply]
can you pls. guide / share document link on how to configure other auth methods for 2fa.

**Reply by Bogey** [reply]
@mjashanks Michael, do you mean that via Google Auth 2fa is possible?

@Manishkparmar80 as an alternative you could put Authelia in front of Budibase.

**Reply by mjashanks** [reply]
I believe that if you use Google Auth, then your Google workspace setup will decide whether to use 2fa or not. I.e. if you have setup 2fa on your google account, then it will apply to Google Auth in Budibase

**Reply by Bogey** [reply]
ok, thanks.
personally I prefer an 'in house' 2fa. the less moving parts, the better imo.

**Comment by Manishkparmar80** [comment]
Error while reviewing preview 
[Error.pdf](https://github.com/Budibase/budibase/files/6562254/Error.pdf)

**Reply by KernOil** [reply]
i have the same problem

**Reply by mike12345567** [reply]
This is a known issue, currently working to resolve it!

**Reply by Manishkparmar80** [reply]
Ok. do let us know once same is gets fixed..

**Comment by Manishkparmar80** [comment]
How i can generate add PDF and Excel report functionality into APP

**Reply by mike12345567** [reply]
Currently PDF export is not a supported feature, this is something that has been mentioned from time to time, if there is a lot of interest I expect this is something we'll look at.

For Excel export it is possible to export a table as a CSV, which can then be imported to Excel if desired.

**Reply by Manishkparmar80** [reply]
Export a table as CSV will not help as the APP user will not have direct access to Table.  May be it will be useful for generation of report from backend.  PDF and Excel report from app is basic required functionality into any of APP.  Output of any APP is MIS report.

**Comment by Manishkparmar80** [comment]
I am getting error while connecting external DB (MS SQL EXPRESS)
i found that resolution for same issue is mentioned at https://stackoverflow.com/questions/55641248/unable-to-connect-to-microsoft-sql-server-using-node-js-mssql-and-express

but not sure about how to do into budibase as the db connection source code access is not available 

![1](https://user-images.githubusercontent.com/84949590/120058216-01d78c80-c067-11eb-9ca3-955b14be1d14.jpg)
![2](https://user-images.githubusercontent.com/84949590/120058220-0603aa00-c067-11eb-86c4-52039b98bdb2.jpg)
![3](https://user-images.githubusercontent.com/84949590/120058221-08fe9a80-c067-11eb-8278-0dece01bd01e.jpg)

**Reply by Manishkparmar80** [reply]
Missed to add error screen shot
![4](https://user-images.githubusercontent.com/84949590/120058256-46632800-c067-11eb-9a27-50153229c45f.jpg)

**Reply by Manishkparmar80** [reply]
Missed to add error screen shot
![4](https://user-images.githubusercontent.com/84949590/120058256-46632800-c067-11eb-9a27-50153229c45f.jpg)

**Reply by shogunpurple** [reply]
@Manishkparmar80 your “server” field looks wrong - try changing it to just ‘localhost’

**Comment by FloMiLe** [comment]
The new release is great news!

Does anybody know whether Budibase works well on a Raspberry Pi, and on which ones?

**Reply by mike12345567** [reply]
Hi @FloMiLe - currently some of the images we use (namely the IBM CouchDB 3.0 image found [here](https://hub.docker.com/r/ibmcom/couchdb3)) do not have ARM image variants right now - this means it would be a little difficult to get it up and running without building some custom images.

We always welcome contributions in this area though, if someone wanted to build a CouchDB 3.0 ARM image and create a docker file which uses all ARM based images then we'd be very happy!

**Reply by FloMiLe** [reply]
@mike12345567, thanks for your reply!

I was not even aware that the CPU architecture could play a role. (Sorry!) I thought anything that can run Docker would also be able to run Budibase. So your input saves me a lot of effort and time...

**Reply by mike12345567** [reply]
> @mike12345567, thanks for your reply!
> 
> I was not even aware that the CPU architecture could play a role. (Sorry!) I thought anything that can run Docker would also be able to run Budibase. So your input saves me a lot of effort and time...

Yes Docker isn't a true virtualisation technology, it passes the hardware resources through directly to each container to share, basically meaning anything running on it needs to support it!

We haven't tested Budibase on any ARM devices, the addition of the CouchDB 3.0 image is very recent (as part of the new full text search functionality we wanted to use) and we felt there wouldn't be a pile of good business use case for running the system on ARM architectures - at least not enough to warrant building and maintaining a custom image for it. If you have a good reason otherwise however we always want to hear it!

**Reply by FloMiLe** [reply]
> If you have a good reason otherwise however we always want to hear it!

I guess my reason is not good enough. It's just that I intend to create Budibase apps for a small group of people. And a Raspberry Pi together with its low energy consumption would be something I can personally afford. (Given that a Raspberry Pi is powerful enough to run Budibase in the first place.)

I will try to find another way to host Budibase...
